public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilderContainer.tsx

```diff
@@ -19,19 +19,28 @@ export interface Props {
 
 /**
  * This component is here just to contain the translation logic between string query and the visual query builder model.
- * @param props
- * @constructor
  */
 export function PromQueryBuilderContainer(props: Props) {
   const { query, onChange, onRunQuery, datasource, data } = props;
+  const [visQuery, setVisQuery] = useState<PromVisualQuery | undefined>();
 
-  const visQuery = buildVisualQueryFromString(query.expr || '').query;
+  useEffect(() => {
+    // Only build the model from expression when editorMode changes
+    const result = buildVisualQueryFromString(query.expr || '');
+    setVisQuery(result.query);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.query.editorMode]);
```

Comment:
This may be problematic. If anything outside changes the expr, we don't reflect that. For example the query patterns right now are outside of the builder so they manipulate the expr directly and so I assume they will stop working after this.

Reply:
There is the same pattern with the labels. You can add a label which starts as totally empty and isn't reflected in the expr at that moment but uses internal state for it and only onBlur or maybe on filling out the whole label onChange get's called. Maybe we can use similar approach here.

Reply:
ah, dam I assumed query patterns only acted on the visual model, but your right. Then I am not sure what do do

Reply:
@aocenas changed the code a bit so that it will rebuild the visual model if expr changes from outside. The other way to solve it is inside OperationEditor but it's not clear how it should know what is a valid restParam and when it should be added to the model vs not.